src/extension.ts

```diff
@@ -183,3 +84,102 @@ function getDashOption(): DashOption {
     exactDocset,
   };
 }
+
+/**
+ * Get docset configuration
+ *
+ * @param {string} languageId e.g javascript, ruby
+ * @return {string}
+ */
+function getDocsets(): string[] {
+  const editor = getEditor();
+  const fileName = path.basename(editor!.document.fileName);
+  const languageId = editor!.document.languageId;
+
+  const fileNameDocsets = getFileNameDocsets(fileName);
+  const languageIdDocsets = getLanguageIdDocsets(languageId);
+
+  // prioritize docset matching by file name then language id
+  return [...fileNameDocsets, ...languageIdDocsets] as string[];
+}
+
+/**
+ * Search in dash for selection syntax documentation
+ */
+function searchSpecific(): void {
+  const editor = getEditor();
+  const query = getSelectedText(editor!);
+  const docsets = getDocsets();
+
+  const dash = new Dash(OS, getDashOption());
+
+  exec(dash.getCommand(query, docsets));
+}
+
+/**
+ * Search in dash for all documentation
+ */
+function searchAll(): void {
+  const editor = getEditor();
+  const query = getSelectedText(editor!);
+  const dash = new Dash(OS, getDashOption());
+
+  exec(dash.getCommand(query));
+}
+
+/**
+ * Search in dash for editor syntax documentation
+ */
+function searchEmptySyntax(): void {
+  const query = '';
+  const docsets = getDocsets();
+  const dash = new Dash(OS, getDashOption());
+
+  exec(dash.getCommand(query, docsets));
+}
+
+/**
+ * Search in dash for editor syntax documentation with a custom query
+ */
+function searchCustomWithSyntax(): void {
+  const docsets = getDocsets();
+  const dash = new Dash(OS, getDashOption());
+
+  const inputOptions: InputBoxOptions = {
+    placeHolder: 'Something to search in Dash.',
+    prompt: 'Enter something to search for in Dash.',
+  };
+
+  window.showInputBox(inputOptions).then(query => {
+    if (query) {
+      // If they actually input code
+      exec(dash.getCommand(query, docsets)); // Open it in dash
+    }
+  });
+}
+
+export function activate(context: ExtensionContext): void {

```

Comment:
what is the reason we put exported function below? I believe exported functions is more important so we want to see it at the beginning of file comparing to non-exported functions

Reply:
Eslint will throw an error https://eslint.org/docs/rules/no-use-before-define. i believe it is best practice too. if you prefer other way, i dont mind either. Do you want me to revert back?

Reply:
interesting. I believe my argument is based on the stepdown rule of Uncle Bob's Clean Code principles. More reference at https://dzone.com/articles/the-stepdown-rule

Maybe we can ignore this rule temporarily.

Reply:
👍